sympy/core/tests/test_operations.py

```diff
@@ -64,3 +65,69 @@ class MyAssoc(AssocOp):
     assert v.args == (u, d)
     # like Add, any unevaluated outer call will flatten inner args
     assert MyAssoc(a, v).args == (a, b, c, d)
+
+def test_add_dispatcher():
+
+    class NewBase1(Expr):
+        @staticmethod
+        def _add_handler(*args, **kwargs):
+            return NewAdd1(*args, **kwargs)
+    class NewAdd1(NewBase1, Add):
+        pass
+    class NewBase2(Expr):
+        _op_priority = 11
+        @staticmethod
+        def _add_handler(*args, **kwargs):
+            return NewAdd2(*args, **kwargs)
+    class NewAdd2(NewBase2, Add):
+        pass
+    a, b, c = Symbol('a'), NewBase1(), NewBase2()
+
+    @add.register_priority(Expr, NewBase1)
+    @add.register_priority(NewBase1, NewBase1)
+    def _(_1, _2):
+        return NewBase1
+
+    # Add called as fallback
+    assert add(1, 2) == 3*S.One
```

Comment:
Maybe assertion should be called on `isinstance` as well... the `==` operator may change with time.

Reply:
Or even better, `assert type( ... ) == ...`, this would prevent matching of subclasses.

Reply:
What about `add(1,2) == Add(1,2)`?  `add` always calls `Add` here so it is guaranteed that two objects are same.

Reply:
Resolving the conversation.